flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaCommitter.java

```diff
@@ -61,13 +62,25 @@ private void commitTransaction(KafkaCommittable committable) {
                 new FlinkKafkaInternalProducer<>(createKafkaProducerConfig(transactionalId))) {
             producer.resumeTransaction(committable.getProducerId(), committable.getEpoch());
             producer.commitTransaction();
-        } catch (InvalidTxnStateException | ProducerFencedException e) {
-            // That means we have committed this transaction before.
+        } catch (InvalidTxnStateException e) {
             LOG.warn(
-                    "Encountered error {} while recovering transaction {}. "
-                            + "Presumably this transaction has been already committed before",
-                    e,
-                    committable);
+                    "Unable to commit recovered transaction ({}) because it's in invalid state. "
+                            + "Most likely the transaction has been aborted for some reason. Please check Kafka logs for more details.",
+                    committable,
+                    e);
+        } catch (ProducerFencedException e) {
+            LOG.warn(
+                    "Unable to commit recovered transaction ({}) because its producer is already fenced."
+                            + " This means that you either have a different producer with the same '{}' (this is"
+                            + " unlikely with the '{}' as all generated ids are unique and shouldn't be reused)"
+                            + " or recovery took longer than '{}' ({}ms). In both cases this most likely signals data loss,"
+                            + " please consult documentation for more details.",
+                    committable,
+                    ProducerConfig.TRANSACTIONAL_ID_CONFIG,
+                    KafkaSink.class.getSimpleName(),
```

Comment:
Why not just inline "KafkaSink" into the message string?

Reply:
It's more of a personal preference.

IMO from maintenance perspective, this is easier to refactor. Also I find it easier to search for references and usages of KafkaSink. Today's tooling can also search for references in comments though, so the benefits may not be as good as I describe them :)